### PR TITLE
Update `update_post_meta()` code where an order ID is passed to use CRUD

### LINKED ISSUE #142: Replace code which uses `update_post_meta()` updating order meta data
## Description

With Custom Order Tables being worked on by the WC Core team ([public announcement](https://developer.woocommerce.com/2022/01/17/the-plan-for-the-woocommerce-custom-order-table/)), we need to begin to think about updating the Subscriptions Core library to remove any code that is calling WP Post API functions.

In this issue, we're tackling only the uses of `update_post_meta()` where an order ID is passed. Subscription IDs or Product IDs are fine at this stage.

### How to update code?
These uses should be replaced with:

```
$order->update_meta_data( $meta_key, $meta_value );
$order->save();
```

or using a setter function if it exists:
```
$order->set_total();
$object->save();
```

### What code needs updating?
 - [x] [`wcs_set_objects_property()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/2a0654bf0385cc3e5b097943aea5589e843b05d0/includes/wcs-compatibility-functions.php#L198-L218)
   - Can use https://github.com/woocommerce/woocommerce-subscriptions/pull/3814/files for inspiration
 - [x] [`save_old_paypal_meta()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/2a0654bf0385cc3e5b097943aea5589e843b05d0/includes/gateways/paypal/includes/class-wcs-paypal-standard-switcher.php#L169-L170)

### Code/classes to ignore
 - [`maybe_record_subscription_payment()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/2a0654bf0385cc3e5b097943aea5589e843b05d0/includes/class-wc-subscriptions-renewal-order.php#L95-L96)
   - Surrounded by pre wc 3.0 condition
 - [`add_order_meta`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/2a0654bf0385cc3e5b097943aea5589e843b05d0/includes/class-wc-subscriptions-synchroniser.php#L1498)
   - Code/function deprecated in subscriptions 2.0
 - `WC_Subscription_Legacy`
   - This class is only loaded pre wc 3.0
 - [`generate_renewal_orders()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/731d98fa36b7037e96e5e5e4f908e091ba68f735/includes/upgrades/class-wc-subscriptions-upgrader.php#L550-L558)
   - This code is in a function that runs when upgrading to WC Subscriptions 1.2 😂  We don't need to touch this code and should probably remove it.
 - `WCS_Repair_2_0_2`
 - `WCS_Upgrade_1_2`
 - 

**Upgrade scripts** can be ignored at this stage as these might be removed from the next version. See internal discussion here: pdjTHR-Y8-p2

## Product impact

- Does this issue affect WooCommerce Subscriptions? yes
- Does this issue affect WooCommerce Payments? yes

## Additional context

Product thread: pdjTHR-Xp-p2

### PR DESCRIPTION
**Pull request by mattallan**
Fixes #142

## Description

When order data is moved to custom tables, updating order meta using `update_post_meta()` will no longer work. This PR updates all cases of `update_post_meta()` where an Order ID is passed - thankfully there wasn't many cases in our codebase 😅 

## How to test this PR

**Testing `wcs_set_objects_property()` changes**

This function is pretty well covered in unit tests, however to test nothing is broken a simple test would be to process a subscription renewal order and make sure the renewal order has the correct meta data.

1. With an active subscription, process a renewal order
2. Confirm the new renewal order has `_billing_address_1` metadata set

**Testing `save_old_paypal_meta()` changes**

This one was a little trickier to test so I just hacked some of the conditions because the only changes we cared about is the order updating post meta:
1. Edit the `save_old_paypal_meta()` function to pass if conditions using, `true ||`
2. Manually set `$subscriptions` to be `$subscriptions = [ 101823 => wcs_get_subscription( 101823 ) ];`
3. Run the following snippet:

```
add_action( 'admin_footer', function() {
	update_post_meta( 101823, '_paypal_subscription_id', 'S-fake1234' );
	update_post_meta( 101823, '_payment_method', 'paypal' );

	WCS_PayPal_Standard_Switcher::save_old_paypal_meta( 101824, [] );
} );
```

4. Confirm the $order meta is set in the DB

image

## Product impact

- [x] Added changelog entry (or does not apply)
- [x] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref

### CODE REVIEW

**Comment by brucealdridge** [comment]
@mattallan The new unit test you added is currently failing.

**Comment by mattallan** [comment]
> The new unit test you added is currently failing.

Whoops! Thanks for the heads up! I copied the unit test changes from https://github.com/woocommerce/woocommerce-subscriptions/pull/3814/files but didn't realise some of the unit test changes were for `wcs_get_object_property()` which isn't changed in this PR.

I've removed the expected deprecated unit test but kept the test additions.

@brucealdridge this should be good to review again. Thanks

**Comment by mattallan** [comment]
One of the existing tests is failing with the unexpected calculated date comparison (off by 1 second). I'll try run the test again but this failing test shouldn't block this PR

**Comment by mattallan** [comment]
I just pushed up [3d1e6f0](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/3d1e6f087345ed79bdd458a6ece25e43bd31b730) which should help improve test flakiness 🤞

**Comment by brucealdridge** [comment]
> I just pushed up [3d1e6f0](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/3d1e6f087345ed79bdd458a6ece25e43bd31b730) which should help improve test flakiness 🤞

This is issue #157  
I did some research on it and it looks like `assertEquals` has removed the delta parameter in the newer version of phpunit so we will need to use `assertEqualsWithDelta` instead.

**Comment by mattallan** [comment]
> This is issue https://github.com/Automattic/woocommerce-subscriptions-core/issues/157

Oh nice! Apologies I didn't see an issue was already created for this!

I'm happy to take it off this PR and handle it separately to avoid confusion that said, I kinda have an issue with using `assertEqualsWithDelta()` where it can be avoided.

In this instance, the issue with this test sometimes failing and sometimes passing is because we're setting up a subscription's start date with `time() -1 month` and then we're calling a bunch of subscription operations (updating dates and status) and then testing that the end date is `assertEqual` to `time()`.

Depending on how slow the test environment is and how much we set as a delta (i.e. 1-2 seconds), we could still this test failing in the future unless we blow out the delta to 5-10 seconds.

The solution I went with was to use a set/stored version `time()` variable to assert expected date calculations, rather than assuming the operations made during the test will be completed within a given delta 😄

File: includes/gateways/paypal/includes/class-wcs-paypal-standard-switcher.php
Comment (brucealdridge) [inline includes/gateways/paypal/includes/class-wcs-paypal-standard-switcher.php]: I'd be tempted to change this condition to do a early return and remove a level of indentation from the method 
```suggestion
		if ( ! wcs_is_order( $order ) || ! wcs_order_contains_switch( $order ) ) {
		    return;
		}
```

File: tests/unit/test-wcs-compatibility-functions.php
Comment (brucealdridge) [inline tests/unit/test-wcs-compatibility-functions.php]: If these are deprecated can we remove them from the unit tests?
Or move to another function and mark as deprecated?

File: tests/unit/test-class-wc-subscriptions.php
Comment (brucealdridge) [inline tests/unit/test-class-wc-subscriptions.php]: I've tested this by adding a sleep(1); between lines 1141-1142 and 1142-1143 and in both cases they fail

**Review by brucealdridge — APPROVED** [review]
Just a few minor nitpicks

File: tests/unit/test-class-wc-subscriptions.php
Comment (mattallan) [reply tests/unit/test-class-wc-subscriptions.php]: I see 🙈 Putting sleeps in the test still causes the same issues that I was trying to prevent.

Because the subscription doesn't have a next payment date, the new calculated end date is being set to the current time, therefore putting sleeps between these lines causes the expected end date I saved to be in the past and an end date can't be in the past, so it still fails.

I had a solution to this by using dates in the future and setting a next payment date on the subscription (which would be the new expected end date), but I think the original purpose of this test is to make sure any end date is set as a result of updating the subscription status to pending-cancel (not validating the end date per se). My approach would slow the test down by updating more dates etc so I'll revert back to using withDelta and go back to the original code 😅

Comment (mattallan) [reply tests/unit/test-class-wc-subscriptions.php]: For reference, here's the updated test I had:

```
$start_time        = time();
$expected_end_time = strtotime( '+1 month', $start_time );

$subscription->update_dates( [
	'start'        => gmdate( 'Y-m-d H:i:s', $start_time ),
	'next_payment' => gmdate( 'Y-m-d H:i:s', $expected_end_time ),
] );

$subscription->update_status( 'pending-cancel' );

$this->assertEquals( $expected_end_time, $subscription->get_time( 'end' ) );
```

File: includes/gateways/paypal/includes/class-wcs-paypal-standard-switcher.php
Comment (mattallan) [reply includes/gateways/paypal/includes/class-wcs-paypal-standard-switcher.php]: Fixed in 886238ca. Thanks for the suggestion!

File: tests/unit/test-class-wc-subscriptions.php
Comment (mattallan) [reply tests/unit/test-class-wc-subscriptions.php]: Fixed in [a218e45](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/a218e4549b1017bce6c82aecfccdc78d03ace860)

File: tests/unit/test-wcs-compatibility-functions.php
Comment (mattallan) [reply tests/unit/test-wcs-compatibility-functions.php]: This param isn't deprecated (copy/paste fail from original PR made to WC Subscriptions). I've removed this wording from the comment to avoid confusion: [4c72ed4](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/4c72ed41d623c4ed5f43f3980e845b24b07f0f6e)

**Comment by mattallan** [comment]
Thanks for the review @brucealdridge . I've addressed the issues you've raised. Since you've approved this PR i'll merge it when everything is green :+1:

Fixes #157
^ is fixed by [a218e45](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/a218e4549b1017bce6c82aecfccdc78d03ace860)

**Pull request merged**
